### PR TITLE
Ability to add contacts to zones and records.

### PR DESCRIPTION
**Pull request by ThomasADavis**
This adds the ability to assign contacts to Zones and DNS Records.

A sample Ansible task to do the assignment is included.

I didn't create any tests, since I haven't had the time to do so..

Currently the netbox web interfaces for Records/Zones doesn't show the contacts - I kept getting either nothing or errors.

https://github.com/netbox-community/netbox/pull/16231 was the reference for this.

### CODE REVIEW

**Comment by peteeckel** [comment]
Hi @ThomasADavis, thanks a lot for this, I think it's a useful addition to the plugin's functionality. 

Unfortunately I won't be able to use it immediately. Main problem is that there already is a `Contact` model in NetBox DNS, which creates all kinds of trouble (already, although that's well-hidden - especially Strawberry doesn't like it at all) and will like cause more trouble once your change gets integrated ... it's been on my list to rename it to `RegistrationContract` for a while now, but I'm not sure if I'll manage it before 1.1.0 gets released.

But there is another problem that is more severe as well: 
Screenshot 2024-08-27 at 09 03 40

I really hate Slack, not least because they use the ransomware model of operation ("Pay **lots** of money or lose your messages after a while - we'll retain them but don't show them to you!"), but unfortunately that's the way NBL choses to communicate with their developers and users ... I took a screen shot to retain the conversation for reference.

https://app.slack.com/client/T01P0D0366B/C03Q7A8C0TS

Bottom line: Someone else tried that before, and I could reproduce it: The contacts pane doesn't show up. We need to get that sorted first, otherwise even adding the functionality to NetBox DNS won't get us anywhere.

Would you like to follow up on this? Getting that sorted out would help at least one other user as well :-)

**Review by peteeckel — COMMENTED** [review]
Hi @ThomasADavis, thanks a lot for this, I think it's a useful addition to the plugin's functionality. 

Unfortunately I won't be able to use it immediately. One problem is that there already is a `Contact` model in NetBox DNS, which creates all kinds of trouble (already, although that's well-hidden - especially Strawberry doesn't like it at all) and will like cause more trouble once your change gets integrated ... it's been on my list to rename it to `RegistrationContract` for a while now, but I'm not sure if I'll manage it before 1.1.0 gets released.

But there is another problem that is more severe as well: 
Screenshot 2024-08-27 at 09 03 40

I really hate Slack, not least because they use the ransomware model of operation ("Pay **lots** of money or lose your messages after a while - we'll retain them but don't show them to you!"), but unfortunately that's the way NBL choses to communicate with their developers and users ... I took a screen shot to retain the conversation for reference.

https://app.slack.com/client/T01P0D0366B/C03Q7A8C0TS

Bottom line: Someone else tried that before, and I could reproduce it: The contacts pane doesn't show up. We need to get that sorted first, otherwise even adding the functionality to NetBox DNS won't get us anywhere.

Would you like to follow up on this? Getting that sorted out would help at least one other user as well :-)

**Comment by peteeckel** [comment]
Hi @ThomasADavis, as you may have been notified about, the first issue is off the table - I prepared a PR that renames the old `Contact` model to `RegistrationContact`. Better to do this with 1.1.0 than to wait for 1.2.0, which won't be around for almost half a year.

When I find the time I'll try to investigate the `ContactsMixin` issue I mentioned above, but currently I'm focussed on some other issues. So If you have some time to spare ...

**Comment by ThomasADavis** [comment]
I was looking at it, but didn't include what I tried because like you found - it doesn't work for some reason.

I have to setup a better develop to work on it, but I will need some time to get to it..  You can see the contact information in the contacts pages, just not in the record/zone pages..

**Comment by peteeckel** [comment]
That's exactly what I found - in principle it works, the data is available and accessible from the API and from scripts, it just doesn't show up in the GUI.

I still think it might have something to do with the templates or the views that the plugin views inherit from. But at the moment I'm a bit time-restricted, so it will have to wait - spent a couple of hours on renaming netbox_dns.models.Contacts (but that was worth it, it was easier than expected).

**Comment by peteeckel** [comment]
@ThomasADavis - by the way, because you are pushing each and every step - you don't need to do that, at least not for running the tests. You can easily run them locally, which is much faster.

Just install `pytest` and `black` in your local development environment, and run `black` and the tests from the command line. 

```
/opt/netbox/netbox/manage.py test --parallel=auto --keepdb netbox_dns
```

If you do migrations in between tests, leave out the `--keepdb`. Some tests will fail unless you stop `netbox-rq` before testing - strictly speaking you don't need it anyway most of the time.

**Comment by ThomasADavis** [comment]
Sorry, that's just me trying to figure out how the tabs work..  I didn't realize it would push against this PR..

The good news is - I'll be trashing this PR, and recreating with a cleaned up version..

The bad news is - till the contacts PR is applied, the contact web page will not work.. it bombs with the with wrong object.

As for the standard mechanism, none of them work as-is; not just contacts, but changelog and journals.

You would have to prove that it doesn't work, and it's not the plugin's fault.

**Comment by ThomasADavis** [comment]
Closing to cleanup the mess I accidentally created..

**Comment by peteeckel** [comment]
Hi @ThomasADavis, it was a good thing you didn't let it rest. I think I solved it, you were absolutely on the right path. Currently some tests fail (I'm not sure if it has anything to do with the change, could also be my test DB that has gone awry), but I managed to add the Contacts tab to the View, NameServer, Zone and Record models (doesn't make much sense for the others IMHO). 

The best thing about it: It works without having to resort to Ansible, the API or custom scripts. You just have to register the right view and url, as I suspected - it just works a bit differently from Journal and ChangeLog.

I'll raise a new PR soon.
